enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -84,155 +85,138 @@ Logging terminology can have overlapping meanings, this document uses the follow
-Future versions of the Elasticsearch store may use `schema` directly.
+*Elasticsearch* outputs MAY specify special index names for records, see below.
 
 #### ClusterLogForwarder configuration
 
-New *pipeline* fields:
+New *pipeline* field:
 
-* `parse`: (string, optional) Legal values are "JSON" or "json" (there may be others in future).
+* `parse`: (string, optional) Only legal value is "json" (there may be others in future).
    If set, attempt to parse log entries as JSON objects.
-* `schemaKey`: (string, optional) Use value of meta-data key as `schema` value, if present.
+
+For each log entry; if `parse: json` is set _and_ the entry is valid JSON, the output record will include a `structured` field _equivalent_ to the JSON entry.
+It may differ in field order and use of white-space.
+
+All record will have a `message` field for backwards compatibility.
+
+#### Output  configuration for type:elasticsearch
+
+New fields for output type `elasticsearch`
+* `indexKey`: (string, optional) Use value of meta-data key as `index` value, if present.
   These keys are supported:
-  - `kubernetes.namespaceName`: Use the namespace name as the schema name.
+  - `kubernetes.namespaceName`: Use the namespace name as the index name.
   - `kubernetes.labels.<key>`: Use the string value of kubernetes label with key `<key>`.
+  - `openshift.labels.<key>`: Use the string value of an openshift label with key `<key>` (see [](forwarder-tagging.md))
   - *other keys may be added in future*
-* `schemaName`: (string, optional)
-  * If `schemaKey` is not set, or the key is missing, use `schemaName` as the `schema` value.
-  * If `schemaKey` is set and the key is present, that takes precedence over `schemaName`
-
-For each log entry; if `parse: json` is set _and_ the entry is valid JSON, the output record will have a `structured` field.
-Otherwise the output record will have a `message` field.
+* `indexName`: (string, optional)
+  * If `indexKey` is not set, or the key is missing, use `indexName` as the `index` value.
+  * If `indexKey` is set and the key is present, that takes precedence over `indexName`
 
-The output record will have a `schema` field if `schemaKey` is set and the meta-data key is present, or if `schemaName` is set.
+When the `defaultOutput` for `elasticsearch`  has an `indexKey` or `indexName` field, the indices will be created automatically on-demand by co-operation between the forwarder and the managed default store.
 
-### User Stories
+#### Default output configuration
 
-#### I want Elasticsearch to index structured logs with an index-per-schema
+To allow index settings on the default elasticsearch output, there is a new section in the ClusterLogForwarder spec:
 
-Setting the `schema` field in outgoing JSON records will send them to different
-Elasticsearch indices. See the [Elasticsearch notes](#es-notes) for more details.
+* outputDefaults (type map, optional):
```

Comment:
I guess we only want to support this for when people use JSON right? cc/ @jcantrill

Reply:
I suggest we implement it for all output types. It is as simple as saving a JSON snippet per output type and applying it before applying config from the API. It is potentially useful in its own right (e.g. I want to create 50 syslog outputs all with the same config except for the target URL, so I create a default output and 50 syslog outputs with just a URL) 
I think it would be harder to clearly document a restricted version of this (and deal with all the users who are surprised when it doesn't work outside those restrictions) than to just implement it properly.